src/mpid/ch4/shm/posix/eager/fbox/fbox_types.h

```diff
@@ -14,18 +14,18 @@
 #include <mpidimpl.h>
 #include "mpidu_init_shm.h"
 
-#define MPIDI_POSIX_FBOX_DATA_LEN  (16 * 1024 - sizeof(uint64_t) - 2 * sizeof(int))
-#define MPIDI_POSIX_FBOX_THRESHOLD (MPIDI_POSIX_FBOX_DATA_LEN)
+#define MPIDI_POSIX_FBOX_SIZE       16 * 1024
 
 typedef struct {
     MPL_atomic_int_t data_ready;
     int is_header;
     size_t payload_sz;
-
-    uint8_t payload[MPIDI_POSIX_FBOX_DATA_LEN];
-
+    uint8_t payload[];
```

Comment:
I may not understand enough about this, but wouldn't this be more likely to cause cache misses? The advantage to having things the other way was that you could be reasonably sure that `payload` was next to `payload_sz`. With a variable length array, you don't know where it's going to be allocated.

Reply:
With variable length arrays, they are always allocated explicitly like this:
https://github.com/pmodels/mpich/blob/82e1145e3d10e2e8ecdaad9c02d4dfc6f510d5df/src/mpid/ch4/shm/posix/eager/fbox/fbox_init.c#L64-L69
So there is not uncertainty on the total sizes that we wanted. And they are always directly accessed through `fbox->payload` so the language guarantee the consistency. I.e. we know exactly where that array is :) .